userspace/libsinsp/user_event.cpp

```diff
@@ -20,6 +20,97 @@ along with sysdig.  If not, see <http://www.gnu.org/licenses/>.
 #include "sinsp_int.h"
 #include "user_event.h"
 
+//
+// event_scope
+//
+
+const std::string event_scope::SCOPE_OP_AND = "and";
+
+// these string lists contain reserved strings; some of the reserved
+// strings are escaped and mandatory to be first in RESERVED_STRINGS
+// and have their escaped counterparts in the REPLACEMENT_STRINGS,
+// in the same order as they appear in RESERVED_STRINGS
+const event_scope::string_list_t event_scope::RESERVED_STRINGS =
+	{"'"/*!must be first!*/, "=", " and ", " "};
+const event_scope::string_list_t event_scope::REPLACEMENT_STRINGS =
+	{"\\'"/*!must be first!*/};
+
+event_scope::event_scope(const std::string& key, const std::string& value)
+{
+	add(key, value, "");
+}
+
+bool event_scope::add(const std::string& key, const std::string& value, const std::string& op)
+{
+	if(check(key))
+	{
+		std::string k(key);
+		std::string o(!m_scope.empty() ? op : "");
+		std::string v(value);
+		replace(v);
+		if(!v.empty())
+		{
+			if(!o.empty())
+			{
+				m_scope.append(1, ' ').append(trim(o)).append(1, ' ');
+			}
+			m_scope.append(trim(k)).append("='").append(trim(v)).append(1, 0x27);
+			return true;
+		}
+	}
+	else
+	{
+		g_logger.log("Scope key has invalid value: [" + key + "], not added to scope.", sinsp_logger::SEV_WARNING);
+	}
+	return false;
+}
+
+string& event_scope::replace(std::string& value)
+{
+	ASSERT(RESERVED_STRINGS.size() >= REPLACEMENT_STRINGS.size());
+
+	if(check(value))
+	{
+		trim(value);
+		string_list_t::const_iterator res_it = RESERVED_STRINGS.cbegin();
+		string_list_t::const_iterator res_end = RESERVED_STRINGS.cend();
+		string_list_t::const_iterator rep_it = REPLACEMENT_STRINGS.cbegin();
+		string_list_t::const_iterator rep_end = REPLACEMENT_STRINGS.cend();
+		for(; res_it != res_end && rep_it != rep_end; ++res_it)
+		{
+			replace_in_place(value, *res_it, *rep_it);
+			if(++rep_it == rep_end) { break; }
+		}
+	}
+	else
+	{
+		g_logger.log("Cannot replace invalid value: [" + value + "], string will be cleared.", sinsp_logger::SEV_WARNING);
+		value.clear();
+	}
+	return value;
+}
+
+// utility function to check that scope entry is valid;
+// valid entries can not contain '=' character or " and " string
+bool event_scope::check(const std::string& scope)
+{
+	if(scope.empty()) { return true; }
+	string_list_t::const_iterator rs = RESERVED_STRINGS.cbegin();
+	++rs;
```

Comment:
This looks pretty sloppy, we'll forget easily to replace this `++` with `+=2` if we'll add reserved_strings in the future.